### DISCUSSION TITLE
Typescript issue with Vanilla/Bound store while testing

### DISCUSSION THREAD

**Discussion by holgergp** [pmndrs/zustand#1564] [Q&A] [answered]
Hi all!
I am struggling to wrap my head around TypeScript types in the context of VanillaStore/Bound store.
My idea is to let test access a vanilla store, while components can use a bound store.

Following the docs I created the following

Create a vanilla store:
```typescript
export const initialState = { foo: [] };
export const fooStore = createStore<FooState>((set, getState) => ({
  ...initialState,
  update: (a) => set((state) => ({ foo: [...state.foo, a] }))
}));
```

Create a hook to expose the vanilla store to the react component:

```typescript
export const useBoundStore: UseBoundStore<StoreApi<FooState>> = (
  selector,
  equals
) => useStore(fooStore, selector, equals);
```
Now I receive a error coming from TypeScript, which I cannot get my head around:

```
Type '<U>(selector: (state: FooState) => U, equals: ((a: U, b: U) => boolean) | undefined) => U' is not assignable to type 'UseBoundStore<StoreApi<FooState>>'.
  Type '<U>(selector: (state: FooState) => U, equals: ((a: U, b: U) => boolean) | undefined) => U' is not assignable to type '{ (): FooState; <U>(selector: (state: FooState) => U, equals?: ((a: U, b: U) => boolean) | undefined): U; }'.
 ```
You can find a running example of this in this codesandbox: https://codesandbox.io/s/zustandvanillaslashboundstore-rrvitm?file=/src/MyStore.ts

What is going on there? What am I missing? Do I use the wrong types? Am I doing this too complicated/not idiomatic?

Any pointer is greatly appreciated! 

Thanks for your awesome work!

Cheers from Düsseldorf

Holger

**Comment by dai-shi** [comment, marked as answer]
You don't need to use `UseBoundStore` type. It's for something else.

Just typing your function should work, like this:
```ts
export const useBoundStore = <T>(
  selector: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
) => useStore(fooStore, selector, equals);
```

@sewera @chrisk-7777 @dbritto-dev This would be a good one for docs if we don't have it yet.

**Reply by holgergp** [reply]
Thank you very much! That works like a charm! 
And thank you for your fast response!

**Reply by holgergp** [reply]
I was looking in the docs for this, but could not find it. I would agree that this might fit nicely in the docs.
If you folks haven't got time for this, I could (try to) make a doc PR for this.

**Reply by chrisk-7777** [reply]
I haven't had a look at the docs for this just yet, but more than happy for you to. 
Please see this comment about running the docs locally: https://github.com/pmndrs/zustand/discussions/1033#discussioncomment-3967766

Once you're done I'm more than happy to have a look over. If you get stuck with local setup, or have any questions, just ping me on that other discussion thread (1033)

**Reply by holgergp** [reply]
Thank you!
I probably won't do it today. But I'll have a try tomorrow.

**Reply by churichard** [reply]
Hi @dai-shi - I'm trying to use this type, but it unfortunately doesn't support using the Persist middleware API (e.g. `useBoundStore.persist.clearStorage()`. Is there a way I can change the typing in order to support this?

**Reply by dai-shi** [reply]
I think you just can/should `fooStore.persist.clearStorage`. No need to attach the store api to the hook, and it's extremely tricky to type it.

**Comment by holgergp** [comment]
I have to reopen this again. While your answer works, there is a tiny issue for me left.
I can use your suggested bound store hook, if I provide a selector function.
```
export const useBoundStore = <T>(
  selector: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
) => useStore(fooStore, selector, equals);
```
If I want to use the hook without providing a selector function, I get 
`Expected 1-2 arguments, but got 0.ts(2554)`

What I can do is provide a second hook, that does not require a selector, or provide a default in the `useBoundStore`-hook:
```
export const useBoundStore = <T>(
  selector: (state: FooState) => T = (state) => state as T,
  equals?: (a: T, b: T) => boolean
) => useStore(fooStore, selector, equals);
```
This does the job, but I get the feeling, that I assume too much of the inner workings of zustand.

Please have a look at the sample here: https://codesandbox.io/s/zustandvanillaslashboundstore-rrvitm?file=/src/MyStore.ts:367-538

What do you think @dai-shi @chrisk-7777?

Thank you!

**Reply by dai-shi** [reply]
Well, if you need to support omitting the selector, you need to overload your function.
```ts
export function useFooStore(): FooState
export function useFooStore<T>(
  selector: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
): T
export function useFooStore<T>(
  selector: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
) {
  return useStore(fooStore, selector, equals)
}
```

**Reply by holgergp** [reply]
Oh! I just learned about function overloading in typescript! (Embarassing! 😓 ) Thanks for that! 😀 
But this one yields:
`This overload signature is not compatible with its implementation signature.ts(2394)`
Using 
```ts
export function useFooStore(): FooState
export function useFooStore<T>(
  selector?: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
): T
export function useFooStore<T>(
  selector?: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
) {
  return useStore(fooStore, selector, equals)
}
```
we fix the above error, but end up not being able to use `useStore`
`undefined' is not assignable to parameter of type '(state: FooState) => T'.
  Type 'undefined' is not assignable to type '(state: FooState) => T`
  
Which we could fix by providing the default implementation from above. But then again we would not need an overload.

Please have a look here: https://codesandbox.io/s/zustandvanillaslashboundstore-rrvitm?file=/src/MyStore.ts

**Reply by dai-shi** [reply]
Oops, nice catch.
Hm, I only know the solution with type assertion: https://codesandbox.io/s/zustandvanillaslashboundstore-forked-ds0vo8

Or we could do this:
```ts
export function useFooStore(): FooState
export function useFooStore<T>(
  selector: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
): T
export function useFooStore<T>(
  selector?: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
) {
  return selector ? useStore(fooStore, selector, equals) : useStore(fooStore)
}
```
But not nice to have extra JS code for TS issue.

**Reply by holgergp** [reply]
But wouldn't the conditional violate the rule of hooks?

**Reply by holgergp** [reply]
But I agree: Having extra conditonals would not be that nice.
But I would not be too happy with that NonNullable-Assertion.

What is your take on having the default value for the selector in the docs?

**Reply by dai-shi** [reply]
> But wouldn't the conditional violate the rule of hooks?

Oh, you are right. That was a bad one.

> But I would not be too happy with that NonNullable-Assertion.

How about this?
```ts
export function useFooStore(): FooState
export function useFooStore<T>(selector: (state: FooState) => T, equals?: (a: T, b: T) => boolean): T
export function useFooStore<T>(selector?: any, equals?: any) {
  return useStore(fooStore, selector, equals)
}
```
We could also do something like this:
```ts
export const useFooStore: {
  (): FooState
  <T>(selector: (state: FooState) => T, equals?: (a: T, b: T) => boolean): T
} = (selector?: any, equals?: any) => useStore(fooStore, selector, equals)
```

> What is your take on having the default value for the selector in the docs?

Do you mean, instead of teaching function overload in #1565, suggest something like this?

```ts
const state = useFooStore((state) => state)
```
Maybe not. You can try your idea in codesandbox and share it.

**Reply by holgergp** [reply]
Ah those are nice! Thank you!

I tried those in https://codesandbox.io/s/zustandvanillaslashboundstore24012023-i2bgu3?file=/src/MyStore.ts:1097-1111

**Ex1:** 
```ts
export function useFooStore(): FooState
export function useFooStore<T>(selector: (state: FooState) => T, equals?: (a: T, b: T) => boolean): T
export function useFooStore<T>(selector?: any, equals?: any) {
  return useStore(fooStore, selector, equals)
}
```
Works! A little wordy "just" to use typescript. But teaches function overloading. Could be a good or a bad thing.

**Ex2:**
```ts
export const useFooStoreEx2: {
  (): FooState;
  <T>(selector: (state: FooState) => T, equals?: (a: T, b: T) => boolean): T;
} = (selector?: any, equals?: any) => useStore(fooStore, selector, equals);
````
I like that one. That is basicially the UseBoundStore type spelled out, right? But sadly this does not compile (like the initial example using UseBoundStoreˆˆ)
```
//Type '(selector?: any, equals?: any) => unknown' is not assignable to type '{ (): FooState; <T>(selector: (state: FooState) => T, equals?: ((a: T, b: T) => boolean) | undefined): T; }'.   Type 'unknown' is not assignable to type 'FooState'.
```
 **Ex3**
 This is the example using a default value for the selector:
 ```ts
export const useFooStoreEx3 = <T>(
  selector: (state: FooState) => T = (state) => state as T,
  equals?: (a: T, b: T) => boolean
) => useStore(fooStore, selector, equals);
```
I like that one, because it is short and "understandable". But it introduces additional JS, just for using TS.

So right now I think it is down to Ex1 and Ex3. I would vote for Ex3. But I would also be happy with Ex1. What do you think?

**Reply by dai-shi** [reply]
For **Ex2**, this should be the fix.

```ts
export const useFooStoreEx2: {
  (): FooState;
  <T>(selector: (state: FooState) => T, equals?: (a: T, b: T) => boolean): T;
} = (selector?: any, equals?: any) => useStore(fooStore, selector, equals) as any;
```

**Reply by dai-shi** [reply]
**Ex0**:
```ts
export const useFooStoreEx0 = <T>(
  selector: (state: FooState) => T,
  equals?: (a: T, b: T) => boolean
) => useStore(fooStore, selector, equals);
```

and use it like `useFooStoreEx0((state) => state)` if we want the entire state.

**Reply by dai-shi** [reply]
What do you think? @chrisk-7777 @sewera @dbritto-dev

**Reply by holgergp** [reply]
What about this (thinking about Ex2):
```
type BoundStoreSelector<U> = {
  (): U;
  <T>(selector: (state: U) => T, equals?: (a: T, b: T) => boolean): T;
};

export const useFooStoreEx2: BoundStoreSelector<FooState> = (
  selector?: any,
  equals?: any
) => useStore(fooStore, selector, equals) as any;
```

The `BoundStoreSelector` type could be something that is provided by zustand, to make it easier to consume for users.

**Reply by dai-shi** [reply]
We don't plan to provide types from the lib which user can create easily. This case falls into it.

**Reply by holgergp** [reply]
I have sandboxed all 4 (Ex0 - Ex3) examples:
https://codesandbox.io/s/zustandvanillaslashboundstore24012023-i2bgu3?file=/src/MyStore.ts

Definition in `MyStore.ts`, Usage in `MyComponent.tsx`

@chrisk-7777 @sewera @dbritto-dev

**Reply by chrisk-7777** [reply]
Took me a while to grok this conversation!
Thank you for providing the sandbox @holgergp , I've been working through the examples while following along in the convo. 

Personally I'm leaning towards the overloaded approach, because I believe it most closely follows how Zustand defines `useStore` itself (https://github.com/pmndrs/zustand/blob/main/src/react.ts#L23)

I understand the downside of teaching what might be a foreign concept (function overloading), but as long as its a documented TS feature I think its appropriate from a documentation perspective to use. Otherwise it becomes a hard line to draw - what TS is considered too advanced and what isn't? 

Even if we include a link to the TS docs - something simple like:

> "Not familiar with this pattern? Checkout the [TS docs on function overloading](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)"

Outside of the teaching side of things, are there any other pitfalls for using the overloaded approach?

**Reply by holgergp** [reply]
Thanks @chrisk-7777!
I would be fine with the overloading solution. Besides that it is a little wordy, I see no other problems with that.
I"ll add that to #1565

**Comment by holgergp** [comment]
I have tried to do a PR:
https://github.com/pmndrs/zustand/pull/1565
@dai-shi @chrisk-7777

**Comment by holgergp** [comment]
Thanks @dai-shi and @chrisk-7777 for your time, great answers and this outcome! 💪

**Comment by dannobytes** [comment]
Related to this discussion, I was trying to [follow documentation](https://docs.pmnd.rs/zustand/guides/typescript#bounded-usestore-hook-for-vanilla-stores) on creating an abstract `createBoundedUseStore` generator b/c I have a need to create bounded store hooks in a more scalable way.

However, the generated bounded hook in the example doesn't get typed correctly.

```ts
const createBoundedUseStore = ((store) => (selector, equals) =>
  useStore(store, selector as any, equals)) as <S extends StoreApi<unknown>>(
  store: S
) => {
  (): ExtractState<S>
  <T>(
    selector?: (state: ExtractState<S>) => T,
    equals?: (a: T, b: T) => boolean
  ): T
}

type ExtractState<S> = S extends { get: () => infer X } ? X : never
```

Has anyone actually gotten this to work as expected? It seems like `ExtractState<S>` and its `infer X` type is not working as intended.

**Reply by dannobytes** [reply]
Oh, I think I just fixed the issue. The store's type is trying to be extracted and inferred from `StoreApi.get()`. However, this function is wrong and doesn't exist on `StoreApi`. The current interface exposes `getState: () => T` instead.

So I'm guessing the docs were either typo'ed and mistaken. Or, `StoreApi.get` was an older interface that was later migrated to `StoreApi.getState`.

For anyone else out there trying to create an abstract bounded use store hook generator, the fix is to replace `get` with `getState` in `ExtractState`.

```ts
type ExtractState<S> = S extends { getState: () => infer X } ? X : never;
```

**Reply by sewera** [reply]
Hi, thanks for spotting the issue! Do you mind contributing to the docs, so that they're up-to-date?

You can find all the necessary info on how to do it in `CONTRIBUTING.md` :)

**Reply by dannobytes** [reply]
You're welcome! And happy to push up a PR to update the docs. I'll get that up today and link to this thread for reference

**Reply by dannobytes** [reply]
https://github.com/pmndrs/zustand/pull/1779

**Comment by dannobytes** [comment]
Ran into one more snag here. When memoizing selectors with custom bound hook, I noticed the selector doesn't get typed properly.

For example:
```ts
const memoizedStuff = useMyStore(useCallback(s => s.getStuff(), []));

// const memoizedStuff: any
```

Any ideas on how to fix the useBoundStore types to have this working?

**Reply by dannobytes** [reply]
Solved. The issue was that the overloaded return functions from `createBoundedUseStore` had marked the first selector as optional. Removing the `?` solved the lost type inference in the memoized callback selector.

Here's the fully fixed version of the abstract bounded store hook generator:
```ts
const createBoundedUseStore = (store => (selector, equals) => useStore(store, selector as never, equals)) as <
  S extends StoreApi<unknown>
>(
  store: S
) => {
  (): ExtractState<S>;
  <U>(selector: (state: ExtractState<S>) => U, equals?: (a: U, b: U) => boolean): U;
};

type ExtractState<S> = S extends { getState: () => infer T } ? T : never;
```
